### ISSUE TITLE
Robust/mature way to minimize joint position steady state error while in rigid contact?

### ISSUE THREAD

**Issue by jeffacce** [facebookresearch/fairo#1368]
## Type of Issue

Select the type of issue:
- [ ] Bug report (to report a bug)
- [ ] Feature request (to request an additional feature)
- [x] Tracker (I am just using this as a tracker)
- [ ] Refactor request
- [ ] Documentation Ask

## Description

_(TLDR: In_ `polymetis`, _is there a robust/proper way to move to a specified joint position with minimal steady state errors, while still being able to handle rigid contacts?)_

In our robotic experiments with the Franka arm, we are using joint position deltas as the robot actions. We found that accurate execution avoids compounding errors and improves rollouts. Our experiments also require that the robot arm interact with rigid objects.

Our current method is adding an integral term to the `move_to_joint_positions` motion generator, and stop updating the integral term if the external torque is too large. This seems a bit suboptimal, and it's not entirely robust upon rigid contact.

## Current Behavior

We have a custom `move_to_joint_positions` function with an integral controller on top of the joint impedance controller to minimize steady state error. This works great when there is no contact, but it exerts too much force when in contact with rigid objects. For example, if the model commands the robot to go into a rigid wall, it will happily push extra hard into the wall as the integral term gradually overcompensates.

## Expected Behavior

Ideally, the controller tries to minimize both steady state error and external torque when interacting with rigid objects.

## Steps to reproduce

Here is our current motion generator (sorry it's a bit messy):
```python
import time
import torch
import bezier
import logging
import polymetis
import numpy as np


def normalize_0_1(lower: float = 0.0, upper: float = 255.0):
    def f(arr):
        arr = arr.clip(lower, upper)
        arr -= arr.min()
        arr /= arr.max()
        return arr

    return f

# motion profile with f', f'', f''' starting and ending at 0
BEZIER_NODES = np.array(
    [
        [-1, 1],
        [-0.5, 1],
        [0, 1],
        [0.45, 1],
        [0.5, 0.5],
        [0.55, 0],
        [1, 0],
        [1.5, 0],
        [2, 0],
    ]
)
BEZIER_NODES[:, 0] = normalize_0_1(-1, 2)(BEZIER_NODES[:, 0])
MOTION_PROFILE = bezier.Curve(BEZIER_NODES.T, degree=len(BEZIER_NODES) - 1)


# linear interpolation: w_a * a + (1 - w_a) * b
def lerp(a: torch.Tensor, b: torch.Tensor, w_a: torch.Tensor):
    w_a = w_a.unsqueeze(1)
    a = a.unsqueeze(0)
    b = b.unsqueeze(0)
    w_b = 1 - w_a
    return w_a * a + w_b * b


class RobotController(polymetis.RobotInterface):
    def move_to_joint_positions_bezier(
        self,
        joint_positions: torch.Tensor,
        time_to_go: float = 4.0,
        fps: int = 60,
        converge_pos_err: bool = False,
        eps: float = 1e-2,
        I_gain: float = 0.1,
        I_clip: float = 0.1,
        torque_limit: float = 5.0,
        torque_comp_gain: float = 0.001,
    ):
        if not hasattr(self, "I_term"):
            self.I_term = torch.zeros(7)
        sleep_period = 1.0 / fps
        n_points = round(time_to_go * fps)
        xs = np.linspace(0, 1, n_points)
        xs, ys = MOTION_PROFILE.evaluate_multi(xs)
        ys = torch.from_numpy(ys)
        start = self.get_joint_positions()
        joint_cmds = lerp(start, joint_positions, ys)
        for i in range(len(joint_cmds)):
            if converge_pos_err:
                cmd = joint_cmds[i] + self.I_term
            else:
                cmd = joint_cmds[i]
            self.update_desired_joint_positions(cmd)
            time.sleep(sleep_period)
        if converge_pos_err:
            # integral controller to drive down steady state error
            err = torch.ones(1)  # position error placeholder
            tau = torch.zeros(7)  # torque placeholder
            while err.norm() > eps and tau.norm() < torque_limit:
                err = self.get_joint_positions() - joint_positions
                tau = torch.Tensor(self.get_robot_state().motor_torques_external)
                logging.info(f"tau: {tau.norm()}")
                self.I_term += -I_gain * err
                self.I_term = torch.clamp(self.I_term, -I_clip, I_clip)
                cmd = joint_positions + self.I_term
                self.update_desired_joint_positions(cmd)
                time.sleep(sleep_period)

    def reset_I_term(self):
        self.I_term = torch.zeros(7)
```

**Comment by exhaustin** [comment]
Hello @jeffacce! We actually pushed an update to Polymetis early August that hugely improves the tracking accuracy. If you haven't upgraded yet, please do so as it might just resolve your issues with tracking accuracy without the need of an integral controller.

If the performance of the updated Polymetis does not meet your specs, you can try increasing the P gains within the [config files](https://github.com/facebookresearch/fairo/blob/main/polymetis/polymetis/conf/robot_client/franka_hardware.yaml). (You can also change them by specifying [Hydra overrides](https://hydra.cc/docs/1.0/advanced/override_grammar/basic/) when running `launch_robot.py`)

If you find that PD alone cannot achieve your accuracy specs while having enough compliance, you might still need the integral term. What I have seen in practice is to decay the integral (multiplying the error integral by a constant slightly smaller than 1 each timestep), or just clamping it to the maximum acceptable force that the integral term can create.
